rfd/0180-sso-as-mfa-method.md

```diff
@@ -0,0 +1,528 @@
++  ALLOW_AS_MFA_METHOD_ONLY = 3;
++}
+
+message OIDCConnectorSpecV3 {
+  ...
++  // AllowAsMFAMethod represents whether this auth connector can be used as an MFA
++  // method or not.
++  AllowAsMFAMethod AllowAsMFAMethod = 19 [(gogoproto.jsontag) = "allow_as_mfa_method,omitempty"];
+}
+
+message SAMLConnectorSpecV2 {
+  ...
++  // AllowAsMFAMethod represents whether this auth connector can be used as an MFA
++  // method or not.
++  AllowAsMFAMethod AllowAsMFAMethod = 17 [(gogoproto.jsontag) = "allow_as_mfa_method,omitempty"];
+}
+ 
+message GithubConnectorSpecV3 {
+  ...
++  // AllowAsMFAMethod represents whether this auth connector can be used as an MFA
++  // method or not.
++  AllowAsMFAMethod AllowAsMFAMethod = 10 [(gogoproto.jsontag) = "allow_as_mfa_method,omitempty"];
+}
+```
+
+**CreatePrivilegedToken**
+```diff
+message OIDCAuthRequest {
+  ...
++  // CreatePrivilegedToken is an option to create a privileged token instead of creating 
++  // a user session. Privileged tokens can be used in place of standard MFA verification for
++  // privileged actions. This action is only allowed if the auth connector is allowed
++  // to be used as an MFA method and if the user is pre-authenticated (not first time login).
++  bool CreatePrivilegedToken = 19 [(gogoproto.jsontag) = "create_privileged_token,omitempty"];
+}
+
+message SAMLAuthRequest {
+  ...
++  // CreatePrivilegedToken is an option to create a privileged token instead of creating 
++  // a user session. Privileged tokens can be used in place of standard MFA verification for
++  // privileged actions. This action is only allowed if the auth connector is allowed
++  // to be used as an MFA method and if the user is pre-authenticated (not first time login).
++  bool CreatePrivilegedToken = 18 [(gogoproto.jsontag) = "create_token,omitempty"];
+}
+
+message GithubAuthRequest {
+  ...
++  // CreatePrivilegedToken is an option to create a privileged token instead of creating 
++  // a user session. Privileged tokens can be used in place of standard MFA verification for
++  // privileged actions. This action is only allowed if the auth connector is allowed
++  // to be used as an MFA method and if the user is pre-authenticated (not first time login).
++  bool CreatePrivilegedToken = 18 [(gogoproto.jsontag) = "create_token,omitempty"];
+}
+```
+
+**AuthConnectorChallenge**
+```diff
+message MFAAuthenticateChallenge {
+  ...
++  // ProviderChallenge is an an auth provider MFA challenge. If set, the client
++  // will attempt to create an auth request with this connector to acquire a
++  // privileged token as a substitute for local MFA.
++  ProviderChallenge ProviderChallenge = 5;
+}
+
++// ProviderChallenge contains auth connector details for completing a provider
++// MFA challenge.
++message ProviderChallenge {
++  // Type is the auth connector type.
++  string Type = 1;
++  // ID is the auth connector ID.
++  string ID = 2;
```

Comment:
Having the id should be enough, no?

Reply:
We need the ID and Type to setup the redirector and create the auth request for the type:
```go
rdConfig, err := tc.SSORedirectorConfig(ctx, connectorID, connectorID, connectorType)
		if err != nil {
			return nil, trace.Wrap(err)
		}

		rdConfig.InitiateSSOLoginFn = func(clientRedirectURL string) (redirectURL string, err error) {
			err = tc.WithRootClusterClient(ctx, func(clt authclient.ClientI) error {
				switch connectorType {
				case constants.OIDC:
					resp, err := clt.CreateOIDCAuthRequest(ctx, types.OIDCAuthRequest{
						Username:              tc.Username,
						ConnectorID:           connectorID,
						Type:                  connectorType,
						CheckUser:             true,
						ClientRedirectURL:     clientRedirectURL,
						CreatePrivilegedToken: true,
					})
					if err != nil {
						return trace.Wrap(err)
					}
					redirectURL = resp.RedirectURL
				case constants.SAML:
					resp, err := clt.CreateSAMLAuthRequest(ctx, types.SAMLAuthRequest{
						Username:              tc.Username,
						ConnectorID:           connectorID,
						Type:                  connectorType,
						ClientRedirectURL:     clientRedirectURL,
						CreatePrivilegedToken: true,
					})
					if err != nil {
						return trace.Wrap(err)
					}
					redirectURL = resp.RedirectURL
				case constants.Github:
					resp, err := clt.CreateGithubAuthRequest(ctx, types.GithubAuthRequest{
						Username:              tc.Username,
						ConnectorID:           connectorID,
						Type:                  connectorType,
						ClientRedirectURL:     clientRedirectURL,
						CreatePrivilegedToken: true,
					})
					if err != nil {
						return trace.Wrap(err)
					}
					redirectURL = resp.RedirectURL
				}
				return nil
			})
			return redirectURL, trace.Wrap(err)
		}
```

Reply:
Thanks for the explanation.

Reply:
FWIW I'm looking into altering this flow to make it more akin to headless authn, making use of the authenticated client to simplify the SSO flow and alleviate some technical challenges, especially with the WebUI.